pkg/iss/pbftsegmentchkp.go

```diff
@@ -0,0 +1,302 @@
+
+		// If the requested Preprepare message is available, send it to the originator of the request.
+		return (&events.EventList{}).PushBack(
+			pbft.eventService.SendMessage(PbftCatchUpResponseSBMessage(preprepare), []t.NodeID{from}),
+		)
+	}
+
+	// If the requested Preprepare message is not available, ignore the request.
+	return &events.EventList{}
+}
+
+// applyMsgCatchUpResponse applies a retransmitted missing committed batch.
+// It only requests hashing of the response,
+// the actual handling of it being performed only when the hash result is available.
+func (pbft *pbftInstance) applyMsgCatchUpResponse(preprepare *isspbftpb.Preprepare, _ t.NodeID) *events.EventList {
+
+	// Ignore preprepare if received in the meantime.
+	// This check is technically redundant, as it is (and must be) performed also after the Preprepare is hashed.
+	// However, it might prevent some unnecessary hash computation if performed here as well.
+	if pbft.slots[pbft.view][t.SeqNr(preprepare.Sn)].Committed {
+		return &events.EventList{}
+	}
+
+	// Request a hash of the received preprepare message.
+	hashRequest := pbft.eventService.HashRequest(
+		[][][]byte{serializePreprepareForHashing(preprepare)},
+		catchUpResponseHashOrigin(preprepare),
+	)
+	return (&events.EventList{}).PushBack(hashRequest)
+}
+
+// applyCatchUpResponseHashResult processes a missing committed batch when its hash becomes available.
+// It is the final step of catching up with an instance-level checkpoint.
+func (pbft *pbftInstance) applyCatchUpResponseHashResult(
+	digest []byte,
+	preprepare *isspbftpb.Preprepare,
+) *events.EventList {
+
+	eventsOut := &events.EventList{}
+
+	// Convenience variables
+	sn := t.SeqNr(preprepare.Sn)
+	slot := pbft.slots[pbft.view][sn]
+
+	// Ignore preprepare if slot is already committed.
+	if pbft.slots[pbft.view][t.SeqNr(preprepare.Sn)].Committed {
+		return &events.EventList{}
+	}
+
+	// Check whether the received batch was actually requested (a faulty node might have sent it on its own).
+	digests := pbft.segmentCheckpoint.Digests()
+	if digests == nil {
+		pbft.logger.Log(logging.LevelWarn, "Ignoring unsolicited CatchUpResponse.", "sn", sn)
+	}
+
+	// Add the missing batch if its digest matches, updating the corresponding Preprepare's view.
+	// Note that copying a Preprepare with an updated view preserves its hash.
+	if bytes.Equal(digests[sn], digest) {
+		slot.catchUp(copyPreprepareToNewView(preprepare, pbft.view), digest)
+	}
+
+	pbft.logger.Log(logging.LevelDebug, "Caught up with segment-level checkpoint.", "sn", sn)
+
+	// If all batches have been committed (i.e. this is the last batch to commit),
+	// send a Done message to all other nodes.
+	// This is required for liveness, see comments for pbftSegmentChkp.
+	if pbft.allCommitted() {
+		eventsOut.PushBackList(pbft.sendDoneMessages())
+	}
+
+	// Deliver batch.
+	eventsOut.PushBack(pbft.eventService.SBEvent(SBDeliverEvent(
+		sn,
+		slot.Preprepare.Batch,
+		slot.Preprepare.Aborted,
+	)))
```

Comment:
Do we want to emit `SBDeliver` event even if the digest does not match?

Reply:
Very good catch! Fiexed.